### PR TITLE
*DRAFT* Initial proposal of Kafka address space

### PR DESCRIPTION
**Pull request by lulf**

### CODE REVIEW

File: documentation/design/proposals/kafka-address-space.adoc [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: false
+    entityOperator:
+      topicOperator: {}
+      userOperator: {}
+----
+
+Creating this resource will not create a Kafka cluster, but it will prepare the `address-space-controller` with the configuration needed to create one when a tenant creates an `AddressSpace` of the `kafka` type. Moreover, settings that are applied on a per-address-space basis will be ignored and set based on the `AddressSpace` definition instead. The following settings are ignored/overridden:
+
```
Comment (k-wall) [inline documentation/design/proposals/kafka-address-space.adoc]: This mean that all Kafka address spaces share the same Kakfa cluster, right?

**Comment by k-wall** [comment]
How would this proposal live alongside #2306 (bridging)

File: documentation/design/proposals/kafka-address-space.adoc (hunk 1 above)
Comment (lulf) [reply documentation/design/proposals/kafka-address-space.adoc]: No, that will come at a later point. I will try to clarify this in the doc, but in the same way we don't create the router and broker when you create the *InfraConfig, we would not create the Kafka cluster when you create the KafkaConfig. The cluster will be created when the AddressSpace is created.

I think longer term, sharing a cluster would be ideal, but it requires modifications to Kafka itself or some proxy in the middle to make it transparent.

**Comment by lulf** [comment]
@k-wall In the first version I think bridging should not be supported. However, we would need to support it at some point, and I imagine it would be in a similar form as for the brokered space, by deploying an amqp-kafka bridge.

File: documentation/design/proposals/kafka-address-space.adoc [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: enmasse.io/v1beta1
+kind: AddressSpace
+metadata:
+  name: myspace
+spec:
+  type: kafka
+  plan: kafka-small
+  endpoints:
```
Comment (ctron) [inline documentation/design/proposals/kafka-address-space.adoc]: Would that include an AMQP 1.0 endpoint?

File: documentation/design/proposals/kafka-address-space.adoc [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+[source,yaml,options="nowrap",subs="attributes"]
+----
+apiVersion: enmasse.io/v1beta1
+kind: Address
+metadata:
+  name: myspace.mytopic
+spec:
+  type: topic
+  plan: kafka-topic-small
+  properties:
```
Comment (ctron) [inline documentation/design/proposals/kafka-address-space.adoc]: What I don't like with the "bag of properties" concept here, is that you no real way to validate that. I would suggest to follow the pattern of Kubernetes instead: create dedicated types for the different choices: see https://github.com/kubernetes/api/blob/master/core/v1/types.go#L49

Comment (lulf) [reply documentation/design/proposals/kafka-address-space.adoc]: @ctron, I’m not sure I understand the reasoning behind distinct types leading to better validation.   

The proposal is not describing a “bag of properties” without a schema. Each field within the properties object may have different and possibly nested types.

This should provide the same level of validation of each property as you would from having separate objects for different address types.

Moreover, any such object would need to allow any property to be omitted and have a default in order to simplify configuration. In addition, the implementation would need to allow unknown properties in order to handle forward and backward compatibility.

With the above requirements I think either direction will give the same benefits. 

A challenge in having separate fields would be that we have the same types (I.e topic) defines in multiple address space types, so i don’t think distinct properties fields for each type makes sense.

If this makes sense to you, I can update the proposal to clarify this bit.

File: documentation/design/proposals/kafka-address-space.adoc (hunk 2 above)
Comment (lulf) [reply documentation/design/proposals/kafka-address-space.adoc]: In the first iteration I think AMQP 1.0 is out of scope. I would like to see this be part of the next iteration, and this design doc should updated when preparing for that.

File: documentation/design/proposals/kafka-address-space.adoc (hunk 3 above)
Comment (lulf) [reply documentation/design/proposals/kafka-address-space.adoc]: So I think the conclusion is to use the `properties` field name, but provide OpenAPI spec to ensure that only a valid set of properties are specified together. I will update this proposal to reflect that, does that sound ok @ctron ?

**Comment by lulf** [comment]
@k-wall Do you want to use this document as a basis for the 'new' design or should I just close this PR?

**Comment by k-wall** [comment]
We should keep the branch so that we can reuse your work later.  It'd be nice to close the PR to keep the PR queue short/clean.

**Pull request closed without merging**